ledger/sandbox/src/main/scala/com/digitalasset/platform/sandbox/stores/ledger/sql/serialisation/ValueHasher.scala

```diff
@@ -0,0 +1,117 @@
+// Copyright (c) 2019 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.digitalasset.platform.sandbox.stores.ledger.sql.serialisation
+
+import java.nio.ByteBuffer
+import java.security.MessageDigest
+
+import com.digitalasset.daml.lf.value.Value
+import com.digitalasset.daml.lf.value.Value.{AbsoluteContractId, VersionedValue}
+
+sealed abstract class HashToken extends Product with Serializable
```

Comment:
I am not convinced we need this abstraction. Although it's an elegant approach to define an ADT for these cases, but are we planning to use these somewhere else too?

Reply:
What would you do? Not use the two step approach, and have the `foldLeft` function directly push bytes to a `MessageDigest` or byte stream?

Reply:
Yes, something like that. And if you need to share code between the path, just have them as small methods. But consider this as a minor comment and leave it as it is if you like. What you have is elegant and easy to understand, it might be a bit more generic though to what we actually need.